### ISSUE TITLE
Implement zeroing of gradients

### ISSUE THREAD

**Issue by jwallwork23** [Cambridge-ICCS/FTorch#336]
Following @jatkinson1000's question about what happens when we run backpropagation on different tensors in https://github.com/Cambridge-ICCS/FTorch/pull/286#discussion_r1998305043, I did some digging.

I think this PyTorch code encapsulates things nicely (print outputs showed as comments):
```python
import torch

a = torch.tensor([1.0], requires_grad=True)
b = torch.tensor([2.0], requires_grad=True)

Q = a * b
print(f"Q = a * b = {Q}")
# Q = a * b = tensor([2.], grad_fn=<MulBackward0>)

P = a * b
print(f"P = a * b = {P}")
# P = a * b = tensor([2.], grad_fn=<MulBackward0>)

Q.backward(torch.ones((1,)), retain_graph=True)
print(f"dQda = {a.grad} ?= b = {b}")
print(f"dQdb = {b.grad} ?= a = {a}")
# dQda = tensor([2.]) ?= b = tensor([2.], requires_grad=True)
# dQdb = tensor([1.]) ?= a = tensor([1.], requires_grad=True)

P.backward(torch.ones((1,)), retain_graph=True)
print(f"dPda = {a.grad} ?= b = {b}")
print(f"dPdb = {b.grad} ?= a = {a}")
# dPda = tensor([4.]) ?= b = tensor([2.], requires_grad=True)
# dPdb = tensor([2.]) ?= a = tensor([1.], requires_grad=True)

R = a * b
print(f"R = a * b = {R}")
# R = a * b = tensor([2.], grad_fn=<MulBackward0>)

R.backward(torch.ones((1,)), retain_graph=True)
print(f"dRda = {a.grad} ?= b = {b}")
print(f"dRdb = {b.grad} ?= a = {a}")
# dRda = tensor([6.]) ?= b = tensor([2.], requires_grad=True)
# dRdb = tensor([3.]) ?= a = tensor([1.], requires_grad=True)

a.grad.zero_()
b.grad.zero_()

S = a * b
print(f"S = a * b = {S}")
# S = a * b = tensor([2.], grad_fn=<MulBackward0>)

S.backward(torch.ones((1,)), retain_graph=True)
print(f"dSda = {a.grad} ?= b = {b}")
print(f"dSdb = {b.grad} ?= a = {a}")
# dSda = tensor([2.]) ?= b = tensor([2.], requires_grad=True)
# dSdb = tensor([1.]) ?= a = tensor([1.], requires_grad=True)
```

In order for the backpropagation to give the correct gradients, we need to zero the gradients beforehand.
